File: ExpressControls/AppButton.vb
' Follow steps 1a or 1b and then 2 to use this custom control in a XAML file.
'
' Step 1a) Using this custom control in a XAML file that exists in the current project.
' Add this XmlNamespace attribute to the root element of the markup file where it is 
' to be used:
'
'     xmlns:MyNamespace="clr-namespace:ExpressControls"
'
'
' Step 1b) Using this custom control in a XAML file that exists in a different project.
' Add this XmlNamespace attribute to the root element of the markup file where it is 
' to be used:
'
'     xmlns:MyNamespace="clr-namespace:ExpressControls;assembly=ExpressControls"
'
' You will also need to add a project reference from the project where the XAML file lives
' to this project and Rebuild to avoid compilation errors:
'
'     Right click on the target project in the Solution Explorer and
'     "Add Reference"->"Projects"->[Browse to and select this project]
'
'
' Step 2)
' Go ahead and use your control in the XAML file. Note that Intellisense in the
' XML editor does not currently work on custom controls and its child elements.
'
'     <MyNamespace:AppButton/>
'

Public Class AppButton
    Inherits Button

    Shared Sub New()
        'This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(AppButton), New FrameworkPropertyMetadata(GetType(AppButton)))
    End Sub

    Public Shared ReadOnly IconProperty As DependencyProperty = DependencyProperty.Register(NameOf(Icon), GetType(Viewbox), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property Icon As Viewbox
        Get
            Return CType(GetValue(IconProperty), Viewbox)
        End Get
        Set(ByVal value As Viewbox)
            SetValue(IconProperty, value)
        End Set
    End Property

    Public Shared ReadOnly TextProperty As DependencyProperty = DependencyProperty.Register(NameOf(Text), GetType(String), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property Text As String
        Get
            Return CStr(GetValue(TextProperty))
        End Get
        Set(ByVal value As String)
            SetValue(TextProperty, value)
        End Set
    End Property

    Public Shared ReadOnly IconVisibilityProperty As DependencyProperty = DependencyProperty.Register(NameOf(IconVisibility), GetType(Visibility), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property IconVisibility As Visibility
        Get
            Return CType(GetValue(IconVisibilityProperty), Visibility)
        End Get
        Set(ByVal value As Visibility)
            SetValue(IconVisibilityProperty, value)
        End Set
    End Property

    Public Shared ReadOnly TextVisibilityProperty As DependencyProperty = DependencyProperty.Register(NameOf(TextVisibility), GetType(Visibility), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property TextVisibility As Visibility
        Get
            Return CType(GetValue(TextVisibilityProperty), Visibility)
        End Get
        Set(ByVal value As Visibility)
            SetValue(TextVisibilityProperty, value)
        End Set
    End Property

    Public Shared ReadOnly MoreVisibilityProperty As DependencyProperty = DependencyProperty.Register(NameOf(MoreVisibility), GetType(Visibility), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property MoreVisibility As Visibility
        Get
            Return CType(GetValue(MoreVisibilityProperty), Visibility)
        End Get
        Set(ByVal value As Visibility)
            SetValue(MoreVisibilityProperty, value)
        End Set
    End Property

    Public Shared ReadOnly GapMarginProperty As DependencyProperty = DependencyProperty.Register(NameOf(GapMargin), GetType(Thickness), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property GapMargin As Thickness
        Get
            Return CType(GetValue(GapMarginProperty), Thickness)
        End Get
        Set(ByVal value As Thickness)
            SetValue(GapMarginProperty, value)
        End Set
    End Property

    Public Shared ReadOnly TextMarginProperty As DependencyProperty = DependencyProperty.Register(NameOf(TextMargin), GetType(Thickness), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property TextMargin As Thickness
        Get
            Return CType(GetValue(TextMarginProperty), Thickness)
        End Get
        Set(ByVal value As Thickness)
            SetValue(TextMarginProperty, value)
        End Set
    End Property

    Public Shared ReadOnly CornerRadiusProperty As DependencyProperty = DependencyProperty.Register(NameOf(CornerRadius), GetType(CornerRadius), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property CornerRadius As CornerRadius
        Get
            Return CType(GetValue(CornerRadiusProperty), CornerRadius)
        End Get
        Set(ByVal value As CornerRadius)
            SetValue(CornerRadiusProperty, value)
        End Set
    End Property

    Public Shared ReadOnly IconSizeProperty As DependencyProperty = DependencyProperty.Register(NameOf(IconSize), GetType(Double), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property IconSize As Double
        Get
            Return CType(GetValue(IconSizeProperty), Double)
        End Get
        Set(ByVal value As Double)
            SetValue(IconSizeProperty, value)
        End Set
    End Property

    Public Shared ReadOnly NoShadowProperty As DependencyProperty = DependencyProperty.Register(NameOf(NoShadow), GetType(Boolean), GetType(AppButton), New FrameworkPropertyMetadata(AddressOf NoShadowPropertyChanged))

    Public Property NoShadow As Boolean
        Get
            Return CBool(GetValue(NoShadowProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(NoShadowProperty, value)
        End Set
    End Property

    Public Shared Sub NoShadowPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim b As Boolean = CBool(e.NewValue)
    End Sub

    Public Shared ReadOnly TextWrappingProperty As DependencyProperty = DependencyProperty.Register(NameOf(TextWrapping), GetType(TextWrapping), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property TextWrapping As TextWrapping
        Get
            Return CType(GetValue(TextWrappingProperty), TextWrapping)
        End Get
        Set(ByVal value As TextWrapping)
            SetValue(TextWrappingProperty, value)
        End Set
    End Property

    Public Shared ReadOnly LastChildFillProperty As DependencyProperty = DependencyProperty.Register(NameOf(LastChildFill), GetType(Boolean), GetType(AppButton), New PropertyMetadata(Nothing))

    Public Property LastChildFill As Boolean
        Get
            Return CBool(GetValue(LastChildFillProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(LastChildFillProperty, value)
        End Set
    End Property

    Public Shared ReadOnly IsUpFacingMenuProperty As DependencyProperty = DependencyProperty.Register(NameOf(IsUpFacingMenu), GetType(Boolean), GetType(AppButton), New FrameworkPropertyMetadata(AddressOf IsUpFacingMenuPropertyChanged))

    Public Property IsUpFacingMenu As Boolean
        Get
            Return CBool(GetValue(IsUpFacingMenuProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(IsUpFacingMenuProperty, value)
        End Set
    End Property

    Public Shared Sub IsUpFacingMenuPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim b As Boolean = CBool(e.NewValue)
    End Sub

End Class

File: ExpressControls/AppButton.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace ExpressControls
{
    /// <summary>
    /// Follow steps 1a or 1b and then 2 to use this custom control in a XAML file.
    ///
    /// Step 1a) Using this custom control in a XAML file that exists in the current project.
    /// Add this XmlNamespace attribute to the root element of the markup file where it is
    /// to be used:
    ///
    ///     xmlns:MyNamespace="clr-namespace:ExpressControls"
    ///
    ///
    /// Step 1b) Using this custom control in a XAML file that exists in a different project.
    /// Add this XmlNamespace attribute to the root element of the markup file where it is
    /// to be used:
    ///
    ///     xmlns:MyNamespace="clr-namespace:ExpressControls;assembly=ExpressControls"
    ///
    /// You will also need to add a project reference from the project where the XAML file lives
    /// to this project and Rebuild to avoid compilation errors:
    ///
    ///     Right click on the target project in the Solution Explorer and
    ///     "Add Reference"->"Projects"->[Browse to and select this project]
    ///
    ///
    /// Step 2)
    /// Go ahead and use your control in the XAML file.
    ///
    ///     <MyNamespace:AppButton/>
    ///
    /// </summary>
    public class AppButton : Button
    {
        static AppButton()
        {
            DefaultStyleKeyProperty.OverrideMetadata(
                typeof(AppButton),
                new FrameworkPropertyMetadata(typeof(AppButton))
            );
        }

        public static readonly DependencyProperty IconProperty = DependencyProperty.Register(
            nameof(Icon),
            typeof(Viewbox),
            typeof(AppButton),
            new PropertyMetadata(null)
        );

        public Viewbox Icon
        {
            get { return (Viewbox)GetValue(IconProperty); }
            set { SetValue(IconProperty, value); }
        }

        public static readonly DependencyProperty TextProperty = DependencyProperty.Register(
            nameof(Text),
            typeof(string),
            typeof(AppButton),
            new PropertyMetadata(null)
        );

        public string Text
        {
            get { return Convert.ToString(GetValue(TextProperty)) ?? ""; }
            set { SetValue(TextProperty, value); }
        }

        public static readonly DependencyProperty IconVisibilityProperty =
            DependencyProperty.Register(
                nameof(IconVisibility),
                typeof(Visibility),
                typeof(AppButton),
                new PropertyMetadata(null)
            );

        public Visibility IconVisibility
        {
            get { return (Visibility)GetValue(IconVisibilityProperty); }
            set { SetValue(IconVisibilityProperty, value); }
        }

        public static readonly DependencyProperty TextVisibilityProperty =
            DependencyProperty.Register(
                nameof(TextVisibility),
                typeof(Visibility),
                typeof(AppButton),
                new PropertyMetadata(null)
            );

        public Visibility TextVisibility
        {
            get { return (Visibility)GetValue(TextVisibilityProperty); }
            set { SetValue(TextVisibilityProperty, value); }
        }

        public static readonly DependencyProperty MoreVisibilityProperty =
            DependencyProperty.Register(
                nameof(MoreVisibility),
                typeof(Visibility),
                typeof(AppButton),
                new PropertyMetadata(null)
            );

        public Visibility MoreVisibility
        {
            get { return (Visibility)GetValue(MoreVisibilityProperty); }
            set { SetValue(MoreVisibilityProperty, value); }
        }

        public static readonly DependencyProperty GapMarginProperty = DependencyProperty.Register(
            nameof(GapMargin),
            typeof(Thickness),
            typeof(AppButton),
            new PropertyMetadata(null)
        );

        public Thickness GapMargin
        {
            get { return (Thickness)GetValue(GapMarginProperty); }
            set { SetValue(GapMarginProperty, value); }
        }

        public static readonly DependencyProperty TextMarginProperty = DependencyProperty.Register(
            nameof(TextMargin),
            typeof(Thickness),
            typeof(AppButton),
            new PropertyMetadata(null)
        );

        public Thickness TextMargin
        {
            get { return (Thickness)GetValue(TextMarginProperty); }
            set { SetValue(TextMarginProperty, value); }
        }

        public static readonly DependencyProperty CornerRadiusProperty =
            DependencyProperty.Register(
                nameof(CornerRadius),
                typeof(CornerRadius),
                typeof(AppButton),
                new PropertyMetadata(null)
            );

        public CornerRadius CornerRadius
        {
            get { return (CornerRadius)GetValue(CornerRadiusProperty); }
            set { SetValue(CornerRadiusProperty, value); }
        }

        public static readonly DependencyProperty IconSizeProperty = DependencyProperty.Register(
            nameof(IconSize),
            typeof(double),
            typeof(AppButton),
            new PropertyMetadata(null)
        );

        public double IconSize
        {
            get { return Convert.ToDouble(GetValue(IconSizeProperty)); }
            set { SetValue(IconSizeProperty, value); }
        }

        public static readonly DependencyProperty NoShadowProperty = DependencyProperty.Register(
            nameof(NoShadow),
            typeof(bool),
            typeof(AppButton),
            new FrameworkPropertyMetadata(NoShadowPropertyChanged)
        );

        public bool NoShadow
        {
            get { return Convert.ToBoolean(GetValue(NoShadowProperty)); }
            set { SetValue(NoShadowProperty, value); }
        }

        public static void NoShadowPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty TextWrappingProperty =
            DependencyProperty.Register(
                nameof(TextWrapping),
                typeof(TextWrapping),
                typeof(AppButton),
                new PropertyMetadata(null)
            );

        public TextWrapping TextWrapping
        {
            get { return (TextWrapping)GetValue(TextWrappingProperty); }
            set { SetValue(TextWrappingProperty, value); }
        }

        public static readonly DependencyProperty LastChildFillProperty =
            DependencyProperty.Register(
                nameof(LastChildFill),
                typeof(bool),
                typeof(AppButton),
                new PropertyMetadata(null)
            );

        public bool LastChildFill
        {
            get { return Convert.ToBoolean(GetValue(LastChildFillProperty)); }
            set { SetValue(LastChildFillProperty, value); }
        }

        public static readonly DependencyProperty IsUpFacingMenuProperty =
            DependencyProperty.Register(
                nameof(IsUpFacingMenu),
                typeof(bool),
                typeof(AppButton),
                new FrameworkPropertyMetadata(IsUpFacingMenuPropertyChanged)
            );

        public bool IsUpFacingMenu
        {
            get { return Convert.ToBoolean(GetValue(IsUpFacingMenuProperty)); }
            set { SetValue(IsUpFacingMenuProperty, value); }
        }

        public static void IsUpFacingMenuPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty IsInTitleBarProperty =
            DependencyProperty.Register(
                nameof(IsInTitleBar),
                typeof(bool),
                typeof(AppButton),
                new FrameworkPropertyMetadata(IsInTitleBarPropertyChanged)
            );

        public bool IsInTitleBar
        {
            get { return Convert.ToBoolean(GetValue(IsInTitleBarProperty)); }
            set { SetValue(IsInTitleBarProperty, value); }
        }

        public static void IsInTitleBarPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty IsCloseBtnProperty = DependencyProperty.Register(
            nameof(IsCloseBtn),
            typeof(bool),
            typeof(AppButton),
            new FrameworkPropertyMetadata(IsCloseBtnPropertyChanged)
        );

        public bool IsCloseBtn
        {
            get { return Convert.ToBoolean(GetValue(IsCloseBtnProperty)); }
            set { SetValue(IsCloseBtnProperty, value); }
        }

        public static void IsCloseBtnPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty HasNoEffectsProperty =
            DependencyProperty.Register(
                nameof(HasNoEffects),
                typeof(bool),
                typeof(AppButton),
                new FrameworkPropertyMetadata(HasNoEffectsPropertyChanged)
            );

        public bool HasNoEffects
        {
            get { return Convert.ToBoolean(GetValue(HasNoEffectsProperty)); }
            set { SetValue(HasNoEffectsProperty, value); }
        }

        public static void HasNoEffectsPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty TextAlignProperty = DependencyProperty.Register(
            nameof(TextAlign),
            typeof(Dock),
            typeof(AppButton),
            new PropertyMetadata(null)
        );

        public Dock TextAlign
        {
            get { return (Dock)GetValue(TextAlignProperty); }
            set { SetValue(TextAlignProperty, value); }
        }
    }
}
